src/irr/core/CFixedCapacityDoublyLinkedList.h

```diff
@@ -0,0 +1,161 @@
+#ifndef __DOUBLY_LINKED_LIST_H_INCLUDED__
+#define __DOUBLY_LINKED_LIST_H_INCLUDED__
+#include "irr/core/alloc/PoolAddressAllocator.h"
+#include "irr/core/Types.h"
+
+namespace irr {
+	namespace core {
+
+		//Struct for use in a doubly linked list. Stores data and pointers to next and previous elements the list, or invalid iterator if it is first/last
+		template<typename Value>
+		struct Snode
+		{
+			_IRR_STATIC_INLINE_CONSTEXPR uint32_t invalid_iterator = DoublyLinkedList<Value>::invalid_iterator;
+
+			Value data;
+			uint32_t prev;
+			uint32_t next;
+
+			Snode() {}
+			Snode(const Value& val) : data(val)
+			{
+				prev = invalid_iterator;
+				next = invalid_iterator;
+			}
+			Snode(Value&& val) : data(std::move(val))
+			{
+				prev = invalid_iterator;
+				next = invalid_iterator;
+			}
+			Snode(Snode<Value>&& other) : data(std::move(other.data)), prev(std::move(other.prev)), next(std::move(other.next))
+			{
+			}
+
+			Snode<Value>& operator=(const Snode<Value>& other)
+			{
+				this->data = other.data;
+				this->prev = other.prev;
+				this->next = other.next;
+				return *this;
+			}
+
+			Snode<Value>& operator=(Snode<Value>&& other)
+			{
+				this->data = std::move(other.data);
+				this->prev = std::move(other.prev);
+				this->next = std::move(other.next);
+				return *this;
+			}
+		};
+
+
+		template<typename Value>
+		class DoublyLinkedList
+		{
+
+			void* reservedSpace;
+			PoolAddressAllocator<uint32_t> alloc;
+			uint32_t m_back;
+			uint32_t m_begin;
+			Snode<Value>* m_array;
+			uint32_t cap;
+
+		public:
+			_IRR_STATIC_INLINE_CONSTEXPR uint32_t invalid_iterator = 0xdeadbeefu;
+
+			//get node at iterator
+			inline Snode<Value>* get(uint32_t address)
+			{
+				return (m_array+address);
+			}
+
+			//remove the last element in the list
+			inline void popBack()
+			{
+				if (m_back == invalid_iterator)	return;
+				auto backNode = getBack();
+				if (backNode->prev != invalid_iterator)
+					get(backNode->prev)->next = invalid_iterator;
+				uint32_t temp = m_back;
+				m_back = backNode->prev;
+				alloc.free_addr(temp, 1u);
```

Comment:
you need to call the destructor on the memory of the back node you've just deallocated.